---
title: "Untitled"
output: html_document
date: '2023-12-13'
---

```{r}

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
if (!require(vegan)) install.packages('vegan')
library(vegan)
if (!require(igraph)) install.packages('igraph')
library(igraph)
if (!require(apcluster)) install.packages('apcluster')
library(apcluster)
if (!require(plyr)) install.packages('plyr')
library(plyr)
if (!require(stringr)) install.packages('stringr')
library(stringr)
if (!require(phyloseq)) install.packages('phyloseq')
library(phyloseq)
if (!require(UpSetR)) install.packages('UpSetR')
library(UpSetR)

```

```{r}
muestras <- c("chile", "tomate" , "maiz")
```

Los análisis generales se harán sobre los OTUs clave que sean bacterias. 

```{r}
rizosfera_chile <- read.csv("../../data/taxonomy/linaje_bacterias_standarizado.de_chile.csv" , header = FALSE , sep = ";" , row.names = 1)
rizosfera_tomate <- read.csv("../../data/taxonomy/linaje_bacterias_standarizado.de_tomate.csv" , header = FALSE , sep = ";" , row.names = 1)
rizosfera_maiz <- read.csv("../../data/taxonomy/linaje_bacterias_standarizado.de_maiz.csv" , header = FALSE , sep = ";" , row.names = 1) 

rizosferas <- list( rizosfera_chile , rizosfera_tomate , rizosfera_maiz )
```


```{r}
red_chile <- read.csv("../../data/networks/chile_species_raw_network.csv")
red_tomate <- read.csv("../../data/networks/tomate_species_raw_network.csv")
red_maiz <- read.csv("../../data/networks/maiz_species_raw_network.csv")

redes <- list(red_chile , red_tomate , red_maiz)
```




```{r}
otus_centrales_chile <- read.csv("../central_otus/reporte_chile1.csv" , row.names = 1 )
otus_centrales_tomate <- read.csv("../central_otus/reporte_tomate1.csv" , row.names = 1 )
otus_centrales_maiz <- read.csv("../central_otus/reporte_maiz1.csv" , row.names = 1 )

otus_centrales <- list(otus_centrales_chile , otus_centrales_tomate , otus_centrales_maiz)
```

```{r}
for (i in 1:3){
  coln_i <- c()
  for (j in 1:(dim(otus_centrales[[i]])[2]-4)) {
    col_j <- make.names(colnames(otus_centrales[[i]])[j])
    coln_i <- c(coln_i , substr(col_j  , 1 ,  nchar(col_j)-21))
  }
  colnames(otus_centrales[[i]]) <- c(coln_i , colnames(otus_centrales[[i]])[(length(colnames(otus_centrales[[i]]))-3):length(colnames(otus_centrales[[i]]))])
  
  }
```



```{r}
data_chile <- read.table("../../data/tables/table.from_chile.txt" , row.names = 1, header = TRUE , sep = "" )
data_tomate <- read.table("../../data/tables/table.from_tomate.txt" , row.names = 1, header = TRUE , sep = "" )
data_maiz <- read.table("../../data/tables/table.from_maiz.txt" , row.names = 1, header = TRUE , sep = "" )

data <- list( data_chile , data_tomate , data_maiz)

```

Poner nombres de muestras

```{r}
for (i in 1:3){
  coln_i <- c()
  for (j in 1:dim(data[[i]])[2]) {
    col_j <- colnames(data[[i]])[j]
    coln_i <- c(coln_i , substr(col_j  , 1 ,  nchar(col_j)-21))
  }
  colnames(data[[i]]) <- coln_i
}
```















##Cotas en las medidas de centralidad para la obtención de OTUs centrales

```{r}
for (i in 1:3){
  print(paste0( "Para rizósfera de " , muestras[i] , " se consideraron OTUs de grado mayor o igual a " , as.character(min(otus_centrales[[i]]$degrees )) , ", cercanía mayor o igual a " , as.character(min(otus_centrales[[i]]$closeness )) , " e intermediación menor o igual a " , as.character(max(otus_centrales[[i]]$betweenness)) , "." ))
}
```

```{r}
taxa_centrales <- list()
for (i in 1:3){
  taxa_centrales[[i]] <- rizosferas[[i]][ intersect( row.names(otus_centrales[[i]]), row.names(rizosferas[[i]]) ) ,]
  print( paste0( "Se obtuvieron "  , length(intersect( row.names(otus_centrales[[i]]), row.names(rizosferas[[i]]) ))  , " OTUs (bacterias) centrales en las muestras de " , muestras[i] ) )
  write.csv( taxa_centrales[[i]] ,  paste0("taxonomy_keystone_otus_", muestras[i] , ".csv"))
  }
```

```{r}
nombres_phyla <- c()
for (i in 1:3){
  nombres_phyla <- union(nombres_phyla ,as.vector(taxa_centrales[[i]][, "V4"]))
  
}

print(nombres_phyla)
```



```{r}
data_phyla <- matrix( nrow = length(nombres_phyla), ncol = length(muestras) )



for (i in 1:3){
  for (j in 1:12) {
    if (is.element(nombres_phyla[j], as.vector(taxa_centrales[[i]][, "V4"] ))){
      data_phyla[j,i] <- 1
    }else
    {
      data_phyla[j,i] <- 0
    }
  }
  
  }


data_phyla <- data.frame(data_phyla , row.names = nombres_phyla )
colnames(data_phyla) <- c(  muestras)
```


```{r}
upset(data_phyla    )
```

```{r}
interseccion <- intersect(as.vector(taxa_centrales[[1]][ , "V4"]) , as.vector(taxa_centrales[[2]][ , "V4"]) )
interseccion <- intersect(interseccion , as.vector(taxa_centrales[[3]][ , "V4"]))
print(interseccion)
print(intersect(as.vector(taxa_centrales[[1]][, "V4"]) , as.vector(taxa_centrales[[2]][ , "V4"]) ))
```


En las siguientes celdas se verifican las intersecciones por pares de los otus clave a nivel familia, género y especie. 

```{r}
nombres_familias <- c()
for (i in 1:3){
  nombres_familias <- union(nombres_familias ,as.vector(taxa_centrales[[i]][, "V7"]))
  
}

print(nombres_familias)
```



```{r}
data_familias <- matrix( nrow = length(nombres_familias), ncol = length(muestras) )



for (i in 1:3){
  for (j in 1:38) {
    if (is.element(nombres_familias[j], as.vector(taxa_centrales[[i]][, "V7"] ))){
      data_familias[j,i] <- 1
    }else
    {
      data_familias[j,i] <- 0
    }
  }
  
  }


data_familias <- data.frame(data_familias , row.names = nombres_familias )
colnames(data_familias) <- c(  muestras)
```


```{r}
upset(data_familias    )
```

A nivel familia los tres tipos de muestras tienen en común son bacterias de la familia Burkholderiaceae (Involvement of Burkholderiaceae and sulfurous volatiles in disease-suppressive soils). 


```{r}
interseccion <- intersect(as.vector(taxa_centrales[[1]][ , "V7"]) , as.vector(taxa_centrales[[2]][ , "V7"]) )
interseccion <- intersect(interseccion , as.vector(taxa_centrales[[3]][ , "V7"]))
print(interseccion)
```


```{r}
nombres_generos <- c()
for (i in 1:3){
  nombres_generos <- union(nombres_generos ,as.vector(taxa_centrales[[i]][, "V8"]))
  
}
print(nombres_generos)
```

```{r}
data_generos <- matrix( nrow = length(nombres_generos), ncol = length(muestras) )



for (i in 1:3){
  for (j in 1:47) {
    if (is.element(nombres_generos[j], as.vector(taxa_centrales[[i]][, "V8"] ))){
      data_generos[j,i] <- 1
    }else
    {
      data_generos[j,i] <- 0
    }
  }
  
  }


data_generos <- data.frame(data_generos , row.names = nombres_generos )
colnames(data_generos) <- c(  muestras)
```


```{r}
upset(data_generos    )
```


Las coincidencias entre los otus centrales a nivel género de las muestras de chile y tomate son:

* Burkholderia (Members of the genus Burkholderia are versatile organisms that occupy a surprisingly wide range of ecological niches. These bacteria are exploited for biocontrol, bioremediation and plant growth promotion purposes,)
* Paenibacillus (Many Paenibacillus species can promote crop growth directly via biological nitrogen fixation, phosphate solubilization, production of the phytohormone indole-3-acetic acid (IAA), and release of siderophores that enable iron acquisition. They can also offer protection against insect herbivores and phytopathogens https://link.springer.com/article/10.1186/s12934-016-0603-7), 
113)


```{r}
print(intersect(as.vector(taxa_centrales[[1]][ , "V8"]) , as.vector(taxa_centrales[[2]][ , "V8"]) ))
```
```{r}
pares <- list(c(1,2) , c(1,3) , c(2,3))
for (i in 1:3){
  print(intersect(as.vector(taxa_centrales[[pares[[i]][1]]][ , "V9"]) , as.vector(taxa_centrales[[pares[[i]][2]]][ , "V9"]) ))
}
```








##OTUs centrales de la rizósfera de chile 

Los OTUs (bacterias) centrales de las muestras de suelo de chile incluyen reguladores de metano y descomponedores de sustancias tóxicas. También hay patógenos humanos. Más detalles se enlistan y verifican a continuación:

* Burkholderia metallica (erosion de metales)
* Advenella mimigardefordensis (able to significantly improve levels of assimilated phosphate)Plant Productivity 
* Sutterella faecalis (heces humanas)
* Eikenella corrodens (cáncer humano?)
* Serratia sp. MYb239 propuesta como especie Serratia rhizosphaerae sp. nov., a novel plant resistance inducer against soft rot disease in tobacco
* Marinobacter (metanotrófica)
* Methylovulum psychrotolerans (metanotrófico, adaptado al frío?)
* Desulfomonile tiedjei (deshalogenante degradador de compuestos tóxicos)
* Paenibacillus chitinolyticus(fijar nitrogeno)
* Nitrosopumilus oxyclinae(solubilización de fosfato)


A continuación se enlistan todos los OTUs centrales con su clasificación a nivel género y nivel especie. 

```{r}
print(taxa_centrales[[1]][ , c("V8" , "V9" )])
```

Solamente entre los OTUs clave de suelo de tomate se recuperó un virus, Spodoptera exigua multiple nucleopolyhedrovirus, que es usado como plaguicida.

```{r}
taxa_virus_chile <- read.csv("../../data/taxonomy/linaje_virus.de_chile.csv" , header = FALSE , sep = ";" )
taxa_virus_chile_centrales <- taxa_virus_chile[ is.element(taxa_virus_chile[,"V1"]  , intersect(as.character(taxa_virus_chile[,"V1"]) , row.names(otus_centrales[[1]]))),]
print(taxa_virus_chile_centrales)
```
```{r}
metadata_chile <- read.csv("../../data/metadata/fastp_metadat.csv" , colClasses = "character")
metadata_chile <- metadata_chile[ which(metadata_chile$Cultivo == "Chile") ,]
metadata_chile <- metadata_chile[ , c("ID" , "Etapa_fenologica" , "Origen")]
for (i in 1:dim(metadata_chile)[1]){
  metadata_chile[i , "ID"] <- make.names(metadata_chile[i , "ID"])
}
metadata_chile <- metadata_chile[which(is.element(metadata_chile[ , "ID"], colnames(otus_centrales[[1]])) ) ,  ]

etapa <- data.frame( ID = metadata_chile[ , "ID"], Phenology =  metadata_chile [ , "Etapa_fenologica"] , Estado = metadata_chile[ , "Origen"],row.names = metadata_chile[ , "ID"])
etapa$Phenology <- c(rep("Developement" , 4) , "Plantation" , NA )
etapa <- sample_data(etapa)
```




```{r}
o_table_key_chile <- otu_table(otus_centrales[[1]][1:(dim(otus_centrales[[1]])[1]-1) , 1:(dim(otus_centrales[[1]])[2]-4) ] , taxa_are_rows = TRUE)  
t_table_key_chile <- tax_table(taxa_centrales[[1]])
row.names(t_table_key_chile@.Data) <- row.names(taxa_centrales[[1]])
o_table_chile <- otu_table(data[[1]], taxa_are_rows = TRUE)
```



```{r}
phy_chile <- phyloseq(otu_table = o_table_chile , sample_data = etapa)
phy_chile <- transform_sample_counts(phy_chile , function(x) x / sum(x) )
meta_ord <- ordinate( phy_chile , method = "PCoA", distance = "bray")
plot_pcoa_muestras <- plot_ordination(physeq = phy_chile, ordination = meta_ord , color = "Phenology")
plot_pcoa_muestras
#ggsave("./chile_muestras_pcoa.png" , plot_pcoa_muestras , device = "png")
```



```{r}
phy_chile_key <- phyloseq(otu_table = o_table_key_chile , tax_table = t_table_key_chile , sample_data = etapa)
phy_chile_key <- transform_sample_counts(phy_chile_key , function(x) x / sum(x) )
meta_ord <- ordinate( phy_chile_key , method = "PCoA", distance = "bray")
plot_pcoa_muestras <- plot_ordination(physeq = phy_chile_key, ordination = meta_ord , color = "Phenology")
plot_pcoa_muestras
#ggsave("./chile__clave_muestras_pcoa.png" , plot_pcoa_muestras , device = "png")
```




```{r}
clave_glom <- tax_glom(phy_chile_key, taxrank = 'ta3')
clave_df <- psmelt(clave_glom)
colnames(clave_df)[9] <- "Phylum"

library(RColorBrewer)
library(ggplot2)
clave_df$Phylum <- as.factor(clave_df$Phylum)
phylum_colors_rel<- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(clave_df$Phylum)))
relative_plot <- ggplot(data=clave_df, aes(x=Sample, y=Abundance, fill=Phylum))+ 
    geom_bar(aes(), stat="identity", position="stack")+
    scale_fill_manual(values = phylum_colors_rel)+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

#absolute_plot | 
relative_plot
#ggsave("phyla_key_chile.png", relative_plot ,device = "png")
```


```{r}
clave_glom <- tax_glom(phy_chile_key, taxrank = 'ta6')
clave_df <- psmelt(clave_glom)

abundance_f <- c()
for (i in 1:length(levels(clave_df$ta6))){
  which_i <- which(as.vector(clave_df$ta6) == levels(clave_df$ta6)[i])
  #print(which_i)
  sum_i <- sum(as.vector(clave_df[which_i,]$Abundance))
  abundance_f <- c(abundance_f , sum_i)
}
familia_abundante <- which(abundance_f > summary(abundance_f)["Median"] )
familia_abundante <- levels(clave_df$ta6)[familia_abundante]

ta6_ab <- c()
for (i in 1:dim(clave_df)[1]){
  if (is.element(as.vector(clave_df$ta6)[i] , familia_abundante )){
    ta6_ab <- c(ta6_ab , as.vector(clave_df$ta6)[i]   )
  } else {
    ta6_ab <- c(ta6_ab , "Less abundant families")
  }
}

clave_df$ta6_ab <- ta6_ab
```


```{r}
library(RColorBrewer)
library(ggplot2)
clave_df$ta6_ab <- as.factor(ta6_ab)
phylum_colors_rel<- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(clave_df$ta6_ab)))
relative_plot <- ggplot(data=clave_df, aes(x=Sample, y=Abundance, fill=ta6_ab))+ 
    geom_bar(aes(), stat="identity", position="stack")+
    scale_fill_manual(values = phylum_colors_rel)
#absolute_plot | 
relative_plot
#ggsave("families_key_chile.png", relative_plot ,device = "png")
```
```{r}
clave_glom <- tax_glom(phy_chile_key, taxrank = 'ta7')
clave_df <- psmelt(clave_glom)

abundance_g <- c()
for (i in 1:length(levels(clave_df$ta7))){
  which_i <- which(as.vector(clave_df$ta7) == levels(clave_df$ta7)[i])
  #print(which_i)
  sum_i <- sum(as.vector(clave_df[which_i,]$Abundance))
  abundance_g <- c(abundance_g , sum_i)
}
genero_abundante <- which(abundance_g > summary(abundance_g)["Median"] )
genero_abundante <- levels(clave_df$ta7)[genero_abundante]

ta7_ab <- c()
for (i in 1:dim(clave_df)[1]){
  if (is.element(as.vector(clave_df$ta7)[i] , genero_abundante )){
    ta7_ab <- c(ta7_ab , as.vector(clave_df$ta7)[i]   )
  } else {
    ta7_ab <- c(ta7_ab , "Less abundant genera")
  }
}

clave_df$ta7_ab <- ta7_ab
```

```{r}
clave_glom <- tax_glom(phy_chile_key, taxrank = 'ta7')
clave_df <- psmelt(clave_glom)
colnames(clave_df)[13] <- "Genus"


phylum_colors_rel<- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(clave_df$Genus)))
relative_plot <- ggplot(data=clave_df, aes(x=Sample, y=Abundance, fill=Genus))+ 
    geom_bar(aes(), stat="identity", position="stack")+
    scale_fill_manual(values = phylum_colors_rel)+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
#absolute_plot | 
relative_plot
#ggsave("genera_key_chile.png", relative_plot ,device = "png")
```

```{r}
key_not_key <- c()
for (i in 1:dim(phy_chile@otu_table@.Data)[1]){
  
  if (is.element( row.names(phy_chile@otu_table@.Data)[i]   , row.names(phy_chile_key@otu_table@.Data)  )   ){
    key_not_key <- c(key_not_key , "key")
  } else {
    key_not_key <- c(key_not_key , "not_key")
  
    }
}
```

```{r}

keys_vs_median <- phy_chile@otu_table@.Data[which(key_not_key == "key") , ]
Medians <- c()
means <- c()
for (i in 1:dim(phy_chile@otu_table@.Data)[2]){
  
  Medians <- c(medians , median(phy_chile@otu_table@.Data[, i]))
  
  }

keys_vs_median <- rbind(Medians , keys_vs_median)
#keys_vs_median <- rbind(keys_vs_median , means)
keys_vs_median <-  psmelt(otu_table(keys_vs_median  , taxa_are_rows = TRUE))
#medians <- psmelt(otu_table(as.data.frame(medians) , taxa_are_rows = FALSE))
```


```{r}
plot <- ggplot(keys_vs_median , aes(x = Sample , y = Abundance , col = OTU , group = OTU))  + geom_line( ) + facet_wrap(~OTU)+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
plot
ggsave("abundance_chile_key_otus_medians.png" , plot ,device = "png")
#ggplot(, aes(x = Sample , y = Abundance , color = OTU)) 
```


```{r}






key_vs_no_key <-  psmelt(phy_chile@otu_table)

key_not_key <- c()
for (i in 1:dim(key_vs_no_key)[1]){
  
  if (is.element( key_vs_no_key[i, "OTU"]   , row.names(phy_chile_key@otu_table@.Data)  )   ){
    key_not_key <- c(key_not_key , "Keystone")
  } else {
    key_not_key <- c(key_not_key , "Not keystone")
  
    }
}
key_vs_no_key <- cbind(key_vs_no_key , key_not_key)
colnames(key_vs_no_key)[4] <- c("Type")
#medians <- psmelt(otu_table(as.data.frame(medians) , taxa_are_rows = FALSE))
```



```{r}
plot <- ggplot(key_vs_no_key , aes(x = Sample , y= Abundance , color = Type  , group = Type ))+ #geom_dotplot(binaxis='y', stackdir='center') +
  stat_summary(fun.data="mean_se", fun.args = list(mult=1), 
                 geom="crossbar", width=0.5) +
 stat_summary(fun="median", fun.args = list(mult=1), 
                 geom="line", width=0.5) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
plot
#  geom_errorbar(aes(ymin=Abundance-sd, ymax=Abundance+sd), width=.2,
 #                position=position_dodge(0.05))
#ggsave("mean_median_key_vs_not_key_chile.png" , plot , device = "png")
```
```{r}
medias <- c()
medianas <- c()
se <- c()
for (i in row.names(taxa_centrales[[1]])){
  medias <- c(medias , mean(phy_chile@otu_table@.Data[i , ] ))
  medianas <- c(medianas , median(phy_chile@otu_table@.Data[i , ] ))
  se <- c(se , mean_se(phy_chile@otu_table@.Data[i , ])[ ,"ymax" ]- mean_se(phy_chile@otu_table@.Data[i , ])[ ,"y" ] )
}
tabla_chile <- data.frame( "OTU" = taxa_centrales[[1]][ , "V9"] ,  "Mean relative abundance" = medias , "Median relative abundance" = medianas, "Standard error" = se  , row.names = row.names(taxa_centrales[[1]]) )
xtable::xtable(tabla_chile , caption = "Keystone OTUs of \textit{Capsicum}" , digits = 8)
```


















```{r}
V(redes[[1]])$label <- NA
V(redes[[1]])$size <- 3
plot(redes[[1]])
```



```{r}
nombres_otus_centrales_chiles <- as.character(otus_centrales[[1]][,"nodos"])
red_otus_centrales_chile <- induced_subgraph(redes[[1]], nombres_otus_centrales_chiles  ,"auto")
print( otus_centrales[[1]][,"nodos"])
print(V(red_otus_centrales_chile))
print(degree( red_otus_centrales_chile))
#V(net_work_centrales)$label <- tax_centrales[,"Rank7"]
```



```{r}
plot(red_otus_centrales_chile)

```


```{r}
otus_no_centrales_chile <- names(V(redes[[1]]))[ is.element(names(V(redes[[1]])), nombres_otus_centrales_chiles) == FALSE]
red_otus_no_centrales_chile <- induced_subgraph(redes[[1]] , otus_no_centrales_chile , "auto" )
plot(red_otus_no_centrales_chile)
```

##Primer reporte de tomate

Aquí se enlistan las especies correspondientes a las muestras de tomate. 

Pseudomonas( Buen colonizador de raíces que se ha postulado para controlar Ralstonia solanacearum  )
Paraburkholderia (fijar nitrogeno  Paraburkholderia spp., are able to improve the macro- and micro-nutrients uptakes by the plants.graminis C4D1M improved tomato seedling re-growth and reduced cell membrane injuries  PG inoculation induced
the expression of five genes (Solyc05g007630.3.1, Solyc07g049700.1.1, Solyc05g013260.3.1,
Solyc09g098100.4.1, Solyc05g005130.3.1) coding for putative late blight resistance protein)
Candidatus Nitrosocosmicus también "centrales" de la canola.
Pseudarthrobacter chlorophenolicus potencial de biofertilizante para tomate (fijar nitrógeno)
Pirellula staleyi( which can degrade putrescine exudate from plant roots visto como importante en rizósfera de granada) 

```{r}
print(taxa_centrales[[2]][ , c( "V8" ,"V9")])
```







```{r}
metadata_tomate <- read.csv("../../data/metadata/fastp_metadat.csv" , colClasses = "character")
metadata_tomate <- metadata_tomate[ which(metadata_tomate$Cultivo == "Tomate") ,]
metadata_tomate <- metadata_tomate[ , c("ID" , "Etapa_fenologica" , "Origen")]
for (i in 1:dim(metadata_tomate)[1]){
  metadata_tomate[i , "ID"] <- make.names(metadata_tomate[i , "ID"])

  }
metadata_tomate <- metadata_tomate[which(is.element(metadata_tomate[ , "ID"], colnames(otus_centrales[[2]])) ) ,  ]

etapa <- data.frame( ID = metadata_tomate[ , "ID"], Phenology =  metadata_tomate [ , "Etapa_fenologica"] , Estado = metadata_tomate[ , "Origen"],row.names = metadata_tomate[ , "ID"])
etapa$Phenology <- c(rep("Developement" , 13) ,"Llenado de fruto"  , rep("Production" , 4))

etapa <- sample_data(etapa)
```



```{r}
o_table_key_tomate <- otu_table(otus_centrales[[2]][ , intersect(row.names(etapa) , colnames(otus_centrales[[2]])) ] , taxa_are_rows = TRUE)  
t_table_key_tomate <- tax_table(taxa_centrales[[2]])
row.names(t_table_key_tomate@.Data) <- row.names(taxa_centrales[[2]])
o_table_tomate <- otu_table(data[[2]][ , intersect(row.names(etapa) , colnames(otus_centrales[[2]])) ], taxa_are_rows = TRUE)
```



```{r}
phy_tomate <- phyloseq(otu_table = o_table_tomate , sample_data = etapa)
phy_tomate <- transform_sample_counts(phy_tomate , function(x) x / sum(x) )
meta_ord <- ordinate( phy_tomate , method = "PCoA", distance = "bray")
plot_pcoa_muestras <- plot_ordination(physeq = phy_tomate, ordination = meta_ord , color = "Phenology")
plot_pcoa_muestras
ggsave("./tomate_muestras_pcoa.png" , plot_pcoa_muestras , device = "png")
```



```{r}
phy_tomate_key <- phyloseq(otu_table = o_table_key_tomate , tax_table = t_table_key_tomate , sample_data = etapa)
phy_tomate_key <- transform_sample_counts(phy_tomate_key , function(x) x / sum(x) )
meta_ord <- ordinate( phy_tomate_key , method = "PCoA", distance = "bray")
plot_pcoa_muestras <- plot_ordination(physeq = phy_tomate_key, ordination = meta_ord , color = "Phenology")
plot_pcoa_muestras
ggsave("./tomate_clave_pcoa.png" , plot_pcoa_muestras , device = "png")
```









```{r}
clave_glom <- tax_glom(phy_tomate_key, taxrank = 'ta3')
clave_df <- psmelt(clave_glom)
colnames(clave_df)[9] <- "Phylum"

library(RColorBrewer)
library(ggplot2)
clave_df$Phylum <- as.factor(clave_df$Phylum)
phylum_colors_rel<- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(clave_df$Phylum)))
relative_plot <- ggplot(data=clave_df, aes(x=Sample, y=Abundance, fill=Phylum))+ 
    geom_bar(aes(), stat="identity", position="stack")+
    scale_fill_manual(values = phylum_colors_rel)+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
#absolute_plot | 
relative_plot
#ggsave("phyla_key_tomate.png", relative_plot ,device = "png")
```

```{r}
clave_glom <- tax_glom(phy_tomate_key, taxrank = 'ta6')
clave_df <- psmelt(clave_glom)
colnames(clave_df)[12] <- "Family"

library(RColorBrewer)
library(ggplot2)
clave_df$Family <- as.factor(clave_df$Family)
phylum_colors_rel<- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(clave_df$Family)))
relative_plot <- ggplot(data=clave_df, aes(x=Sample, y=Abundance, fill=Family))+ 
    geom_bar(aes(), stat="identity", position="stack")+
    scale_fill_manual(values = phylum_colors_rel)+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
#absolute_plot | 
relative_plot
#ggsave("phyla_key_tomate.png", relative_plot ,device = "png")
```


```{r}
clave_glom <- tax_glom(phy_tomate_key, taxrank = 'ta6')
clave_df <- psmelt(clave_glom)


abundance_f <- c()
for (i in 1:length(levels(clave_df$ta6))){
  which_i <- which(as.vector(clave_df$ta6) == levels(clave_df$ta6)[i])
  #print(which_i)
  sum_i <- sum(as.vector(clave_df[which_i,]$Abundance))
  abundance_f <- c(abundance_f , sum_i)
}
familia_abundante <- which(abundance_f > summary(abundance_f)["Median"] )
familia_abundante <- levels(clave_df$ta6)[familia_abundante]

ta6_ab <- c()
for (i in 1:dim(clave_df)[1]){
  if (is.element(as.vector(clave_df$ta6)[i] , familia_abundante )){
    ta6_ab <- c(ta6_ab , as.vector(clave_df$ta6)[i]   )
  } else {
    ta6_ab <- c(ta6_ab , "Less abundant families")
  }
}

clave_df$ta6_ab <- ta6_ab
```


```{r}
library(RColorBrewer)
library(ggplot2)
colnames(clave_df)[13] <- "Family"

clave_df$Family <- as.factor(clave_df$Family)
phylum_colors_rel<- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(clave_df$Family)))
relative_plot <- ggplot(data=clave_df, aes(x=Sample, y=Abundance, fill=Family))+ 
    geom_bar(aes(), stat="identity", position="stack")+
    scale_fill_manual(values = phylum_colors_rel)+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
#absolute_plot | 
relative_plot
ggsave("family_key_tomate.png", relative_plot ,device = "png")
```
```{r}
clave_glom <- tax_glom(phy_tomate_key, taxrank = 'ta7')
clave_df <- psmelt(clave_glom)

abundance_g <- c()
for (i in 1:length(levels(clave_df$ta7))){
  which_i <- which(as.vector(clave_df$ta7) == levels(clave_df$ta7)[i])
  #print(which_i)
  sum_i <- sum(as.vector(clave_df[which_i,]$Abundance))
  abundance_g <- c(abundance_g , sum_i)
}
genero_abundante <- which(abundance_g > summary(abundance_g)["Median"] )
genero_abundante <- levels(clave_df$ta7)[genero_abundante]

ta7_ab <- c()
for (i in 1:dim(clave_df)[1]){
  if (is.element(as.vector(clave_df$ta7)[i] , genero_abundante )){
    ta7_ab <- c(ta7_ab , as.vector(clave_df$ta7)[i]   )
  } else {
    ta7_ab <- c(ta7_ab , "g__")
  }
}

clave_df$ta7_ab <- ta7_ab
```

```{r}
colnames(clave_df)[14] <- "Genus"

clave_df$Genus <- as.factor(clave_df$Genus)
phylum_colors_rel<- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(clave_df$Genus)))
relative_plot <- ggplot(data=clave_df, aes(x=Sample, y=Abundance, fill=Genus))+ 
    geom_bar(aes(), stat="identity", position="stack")+
    scale_fill_manual(values = phylum_colors_rel)+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
#absolute_plot | 
relative_plot
```

```{r}
key_not_key <- c()
for (i in 1:dim(phy_tomate@otu_table@.Data)[1]){
  
  if (is.element( row.names(phy_tomate@otu_table@.Data)[i]   , row.names(phy_tomate_key@otu_table@.Data)  )   ){
    key_not_key <- c(key_not_key , "key")
  } else {
    key_not_key <- c(key_not_key , "not_key")
  
    }
}
```

```{r}

keys_vs_median <- phy_tomate@otu_table@.Data[which(key_not_key == "key") , ]
medians <- c()
means <- c()
for (i in 1:dim(phy_tomate@otu_table@.Data)[2]){
  
  medians <- c(medians , median(phy_tomate@otu_table@.Data[, i]))
  
  }

keys_vs_median <- rbind(keys_vs_median , medians)
#keys_vs_median <- rbind(keys_vs_median , means)
keys_vs_median <-  psmelt(otu_table(keys_vs_median  , taxa_are_rows = TRUE))
```









```{r}
ggplot(keys_vs_median , aes(x = Sample , y = Abundance , col = OTU , group = OTU))  + geom_line( ) + facet_wrap(~OTU)
#ggplot(, aes(x = Sample , y = Abundance , color = OTU)) 
```

```{r}






key_vs_no_key <-  psmelt(phy_tomate@otu_table)

key_not_key <- c()
for (i in 1:dim(key_vs_no_key)[1]){
  
  if (is.element( key_vs_no_key[i, "OTU"]   , row.names(phy_tomate_key@otu_table@.Data)  )   ){
    key_not_key <- c(key_not_key , "Keystone")
  } else {
    key_not_key <- c(key_not_key , "Not keystone")
  
    }
}
key_vs_no_key <- cbind(key_vs_no_key , key_not_key)
colnames(key_vs_no_key)[4] <- c("Type")
#medians <- psmelt(otu_table(as.data.frame(medians) , taxa_are_rows = FALSE))
```



```{r}
plot <- ggplot(key_vs_no_key , aes(x = Sample , y= Abundance , color = Type  , group = Type))+ #geom_dotplot(binaxis='y', stackdir='center') +
  stat_summary(fun.data="mean_se", fun.args = list(mult=1), 
                 geom="crossbar", width=0.5) +
 stat_summary(fun="median", fun.args = list(mult=1), 
                 geom="line", width=0.5)+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
plot

ggsave("mean_median_key_vs_not_key_tomate.png" , plot , device = "png")
#  geom_errorbar(aes(ymin=Abundance-sd, ymax=Abundance+sd), width=.2,
 #                position=position_dodge(0.05))
```

















```{r}
ggplot(keys_vs_median , aes(x = Sample , y = Abundance , color = OTU)) + geom_point()
```



```{r}
key <- which(key_not_key == "key") 
key <- phy_chile@otu_table@.Data[key , ]



cfds <- list()
for (i in 1:25){
  alea <- sample(which(key_not_key == "not_key") , size = 1)
  alea <- phy_chile@otu_table@.Data[alea , ]
  dist <- c(alea , alea_key)
  comp <- c(rep("not_key" , 6), rep("alea_key" , 6) )

  cdfs[[i]] <- data.frame(dist , comp )


}
#for (i in 1:25){
 #ggsave(paste0("comp_distr.png" , as.character(i)) ,  ggplot(cdfs[[i]] , aes(x = comp, y = dist)) + geom_boxplot() , device = "png")
#}


```

```{r}
medias <- c()
medianas <- c()
se <- c()
for (i in row.names(taxa_centrales[[2]])){
  medias <- c(medias , mean(phy_tomate@otu_table@.Data[i , ] ))
  medianas <- c(medianas , median(phy_tomate@otu_table@.Data[i , ] ))
  se <- c(se , mean_se(phy_tomate@otu_table@.Data[i , ])[ ,"ymax" ]- mean_se(phy_tomate@otu_table@.Data[i , ])[ ,"y" ] )
}
tabla_tomate <- data.frame( "OTU" = taxa_centrales[[2]][ , "V8"] ,  "Mean relative abundance" = medias , "Median relative abundance" = medianas, "Standard error" = se  , row.names = row.names(taxa_centrales[[2]]) )
xtable::xtable(tabla_tomate , caption = "Keystone OTUs of \textit{Solanum lycopersicum}" , digits = 8)
```






















En la tabla conseguida en la siguiente celda se observa que los otus centrales están presentes en sus muestras más que la mediana del resto de los otus, si bien menor a la media. Una de las muestras tiene la mayoría de los otus con una sola instancia. 

```{r}
data_tomate <- read.table("../../data/tables/table.from_tomate.txt" , row.names = 1, header = TRUE , sep = "" )

medias <- c()
medianas <- c()
for (i in 1:dim(data_tomate)[2]){
  medias <- c(medias , mean(data_tomate[ , i]))
  medianas <- c(medianas , median(data_tomate[ , i]))
  }

data_tomate <- data_tomate[ row.names(otus_centrales[[2]]),]
data_tomate <- rbind(medias , data_tomate)
data_tomate <- rbind(medianas , data_tomate)

row.names(data_tomate) <- c("Medianas" , "Medias" , row.names(otus_centrales[[2]]))
head(data_tomate)
```

```{r}
V(redes[[2]])$label <- NA
V(redes[[2]])$size <- 3
plot(redes[[2]])
```

```{r}
nombres_otus_centrales_tomate <- as.character(otus_centrales[[2]][,"nodos"])
red_otus_centrales_tomate <- induced_subgraph(redes[[2]], nombres_otus_centrales_tomate  ,"auto")
print( otus_centrales[[2]][,"nodos"])
print(V(red_otus_centrales_tomate))
print(degree( red_otus_centrales_tomate))
#V(net_work_centrales)$label <- tax_centrales[,"Rank7"]
```

```{r}
plot(red_otus_centrales_tomate)
```

```{r}
otus_no_centrales_tomate <- names(V(redes[[2]]))[ is.element(names(V(redes[[2]])), nombres_otus_centrales_tomate) == FALSE]
red_otus_no_centrales_tomate <- induced_subgraph(redes[[2]] , otus_no_centrales_tomate , "auto" )
plot(red_otus_no_centrales_tomate)
```




##Primer reporte de maiz

Aquí se enlistan las especies correspondientes a las muestras de maiz. 

Phoenicibacter congonensis( microbiota intestinal humana )
Bacillus glycinifermentans(solubilizacion de zinc, reportado como estimulante de crecimiento de raíz y tallo en trigo)

Ralstonia solanacearum(patógeno conocido)
Erwinia persicina (produces large amounts of mucilaginous exopolysaccharides on King’s B ciertas cepas mejora de alfalfa , this fungal species is mainly responsible for the accumulation of the mycotoxin disminuye virulence factor deoxynivalenol (DON).)

```{r}
print(taxa_centrales[[3]][ , c( "V8" ,"V9")])
```

En la tabla conseguida en la siguiente celda se observa que los otus centrales están presentes en sus muestras más que la mediana del resto de los otus, si bien menor a la media. Una de las muestras tiene la mayoría de los otus con una sola instancia. 










```{r}
metadata_maiz <- read.csv("../../data/metadata/fastp_metadat.csv" , colClasses = "character")
metadata_maiz <- metadata_maiz[ which(metadata_maiz$Cultivo == "Maiz") ,]
metadata_maiz <- metadata_maiz[ , c("ID" , "Etapa_fenologica" , "Origen")]
for (i in 1:dim(metadata_maiz)[1]){
  metadata_maiz[i , "ID"] <- make.names(metadata_maiz[i , "ID"])
}
metadata_maiz <- metadata_maiz[which(is.element(metadata_maiz[ , "ID"], colnames(otus_centrales[[3]])) ) ,  ]

etapa <- data.frame( ID = metadata_maiz[ , "ID"], Phenology =  metadata_maiz [ , "Etapa_fenologica"] , Estado = metadata_maiz[ , "Origen"],row.names = metadata_maiz[ , "ID"])
etapa$Phenology <- c(rep("Formacion de espiga" , 4) , rep("Production" , 2))
etapa <- sample_data(etapa)
```




```{r}
o_table_key_maiz <- otu_table(otus_centrales[[3]][1:(dim(otus_centrales[[3]])[1]-1) , 1:(dim(otus_centrales[[3]])[2]-4) ] , taxa_are_rows = TRUE)  
t_table_key_maiz <- tax_table(taxa_centrales[[3]])
row.names(t_table_key_maiz@.Data) <- row.names(taxa_centrales[[3]])
o_table_maiz <- otu_table(data[[3]], taxa_are_rows = TRUE)
```



```{r}
library(ggplot2)
phy_maiz <- phyloseq(otu_table = o_table_maiz , sample_data = etapa)
phy_maiz <- transform_sample_counts(phy_maiz , function(x) x / sum(x) )
meta_ord <- ordinate( phy_maiz , method = "PCoA", distance = "bray")
plot_pcoa_muestras <- plot_ordination(physeq = phy_maiz, ordination = meta_ord , color = "Phenology")
plot_pcoa_muestras
#ggsave("maiz_muestras_pcoa.png" , plot_pcoa_muestras , device = "png")
```



```{r}
phy_maiz_key <- phyloseq(otu_table = o_table_key_maiz , tax_table = t_table_key_maiz , sample_data = etapa)
phy_maiz_key <- transform_sample_counts(phy_maiz_key , function(x) x / sum(x) )
meta_ord <- ordinate( phy_maiz_key , method = "PCoA", distance = "bray")
plot_pcoa_muestras <- plot_ordination(physeq = phy_maiz_key, ordination = meta_ord , color = "Phenology")
plot_pcoa_muestras
ggsave("maiz_pcoa_key_otus.png" , plot_pcoa_muestras , device = "png")
```




```{r}
clave_glom <- tax_glom(phy_maiz_key, taxrank = 'ta3')
clave_df <- psmelt(clave_glom)

colnames(clave_df)[9] <- "Phylum"

library(RColorBrewer)
library(ggplot2)
clave_df$Phylum <- as.factor(clave_df$Phylum)
phylum_colors_rel<- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(clave_df$Phylum)))
relative_plot <- ggplot(data=clave_df, aes(x=Sample, y=Abundance, fill=Phylum))+ 
    geom_bar(aes(), stat="identity", position="stack")+
    scale_fill_manual(values = phylum_colors_rel)+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
#absolute_plot | 
relative_plot
#ggsave("phyla_key_maiz.png", relative_plot ,device = "png")
```
```{r}
clave_glom <- tax_glom(phy_maiz_key, taxrank = 'ta7')
clave_df <- psmelt(clave_glom)

colnames(clave_df)[13] <- "Genus"

library(RColorBrewer)
library(ggplot2)
clave_df$Genus <- as.factor(clave_df$Genus)
phylum_colors_rel<- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(clave_df$Genus)))
relative_plot <- ggplot(data=clave_df, aes(x=Sample, y=Abundance, fill=Genus))+ 
    geom_bar(aes(), stat="identity", position="stack")+
    scale_fill_manual(values = phylum_colors_rel)+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
#absolute_plot | 
relative_plot
#ggsave("genera_key_maiz.png", relative_plot ,device = "png")
```
```{r}
key_not_key <- c()
for (i in 1:dim(phy_maiz@otu_table@.Data)[1]){
  
  if (is.element( row.names(phy_maiz@otu_table@.Data)[i]   , row.names(phy_maiz_key@otu_table@.Data)  )   ){
    key_not_key <- c(key_not_key , "key")
  } else {
    key_not_key <- c(key_not_key , "not_key")
  
    }
}
```

```{r}

keys_vs_median <- phy_maiz@otu_table@.Data[which(key_not_key == "key") , ]
Medians <- c()
Means <- c()
for (i in 1:dim(phy_maiz@otu_table@.Data)[2]){
  
  Medians <- c(Medians , median(phy_maiz@otu_table@.Data[, i]))
  Means <- c(Means , mean(phy_maiz@otu_table@.Data[, i]))
  }

keys_vs_median <- rbind(Medians , keys_vs_median)
keys_vs_median <- rbind(keys_vs_median , Means)
keys_vs_median <-  psmelt(otu_table(keys_vs_median  , taxa_are_rows = TRUE))
#medians <- psmelt(otu_table(as.data.frame(medians) , taxa_are_rows = FALSE))
```


```{r}
plot <- ggplot(keys_vs_median , aes(x = Sample , y = Abundance , col = OTU , group = OTU))  + geom_line( ) + facet_wrap(~OTU)+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
plot
ggsave("abundance_maiz_key_otus_medians.png" , plot ,device = "png")
#ggplot(, aes(x = Sample , y = Abundance , color = OTU)) 
```


```{r}






key_vs_no_key <-  psmelt(phy_maiz@otu_table)

key_not_key <- c()
for (i in 1:dim(key_vs_no_key)[1]){
  
  if (is.element( key_vs_no_key[i, "OTU"]   , row.names(phy_maiz_key@otu_table@.Data)  )   ){
    key_not_key <- c(key_not_key , "Keystone")
  } else {
    key_not_key <- c(key_not_key , "Not keystone")
  
    }
}
key_vs_no_key <- cbind(key_vs_no_key , key_not_key)
colnames(key_vs_no_key)[4] <- c("Type")
#medians <- psmelt(otu_table(as.data.frame(medians) , taxa_are_rows = FALSE))
```



```{r}
plot <- ggplot(key_vs_no_key , aes(x = Sample , y= Abundance , color = Type  , group = Type ))+ #geom_dotplot(binaxis='y', stackdir='center') +
  stat_summary(fun.data="mean_se", fun.args = list(mult=1), 
                 geom="crossbar", width=0.5) +
 stat_summary(fun="median", fun.args = list(mult=1), 
                 geom="line", width=0.5) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
plot
#  geom_errorbar(aes(ymin=Abundance-sd, ymax=Abundance+sd), width=.2,
 #                position=position_dodge(0.05))
ggsave("mean_median_key_vs_not_key_maiz.png" , plot , device = "png")
```

```{r}
medias <- c()
medianas <- c()
se <- c()
for (i in row.names(taxa_centrales[[3]])){
  medias <- c(medias , mean(phy_maiz@otu_table@.Data[i , ] ))
  medianas <- c(medianas , median(phy_maiz@otu_table@.Data[i , ] ))
  se <- c(se , mean_se(phy_maiz@otu_table@.Data[i , ])[ ,"ymax" ]- mean_se(phy_maiz@otu_table@.Data[i , ])[ ,"y" ] )
}
tabla_tomate <- data.frame( "OTU" = taxa_centrales[[3]][ , "V"] ,  "Mean relative abundance" = medias , "Median relative abundance" = medianas, "Standard error" = se  , row.names = row.names(taxa_centrales[[3]]) )
xtable::xtable(tabla_tomate , caption = "Keystone OTUs of textit{Zea mays}" , digits = 8)
```















```{r}
clave_glom <- tax_glom(phy_maiz_key, taxrank = 'ta6')
clave_df <- psmelt(clave_glom)

abundance_f <- c()
for (i in 1:length(levels(clave_df$ta6))){
  which_i <- which(as.vector(clave_df$ta6) == levels(clave_df$ta6)[i])
  #print(which_i)
  sum_i <- sum(as.vector(clave_df[which_i,]$Abundance))
  abundance_f <- c(abundance_f , sum_i)
}
familia_abundante <- which(abundance_f > summary(abundance_f)["1st Qu."] )
familia_abundante <- levels(clave_df$ta6)[familia_abundante]

ta6_ab <- c()
for (i in 1:dim(clave_df)[1]){
  if (is.element(as.vector(clave_df$ta6)[i] , familia_abundante )){
    ta6_ab <- c(ta6_ab , as.vector(clave_df$ta6)[i]   )
  } else {
    ta6_ab <- c(ta6_ab , "f__")
  }
}

clave_df$ta6_ab <- ta6_ab
```


```{r}
library(RColorBrewer)
library(ggplot2)
clave_df$ta6_ab <- as.factor(ta6_ab)
phylum_colors_rel<- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(clave_df$ta6_ab)))
relative_plot <- ggplot(data=clave_df, aes(x=Sample, y=Abundance, fill=ta6_ab))+ 
    geom_bar(aes(), stat="identity", position="stack")+
    scale_fill_manual(values = phylum_colors_rel)
#absolute_plot | 
relative_plot
```
```{r}
clave_glom <- tax_glom(phy_maiz_key, taxrank = 'ta7')
clave_df <- psmelt(clave_glom)

abundance_g <- c()
for (i in 1:length(levels(clave_df$ta7))){
  which_i <- which(as.vector(clave_df$ta7) == levels(clave_df$ta7)[i])
  #print(which_i)
  sum_i <- sum(as.vector(clave_df[which_i,]$Abundance))
  abundance_g <- c(abundance_g , sum_i)
}
genero_abundante <- which(abundance_g > summary(abundance_g)["1st Qu."] )
genero_abundante <- levels(clave_df$ta7)[genero_abundante]

ta7_ab <- c()
for (i in 1:dim(clave_df)[1]){
  if (is.element(as.vector(clave_df$ta7)[i] , genero_abundante )){
    ta7_ab <- c(ta7_ab , as.vector(clave_df$ta7)[i]   )
  } else {
    ta7_ab <- c(ta7_ab , "g__")
  }
}

clave_df$ta7_ab <- ta7_ab
```

```{r}
clave_df$ta7_ab <- as.factor(ta7_ab)
phylum_colors_rel<- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(clave_df$ta7_ab)))
relative_plot <- ggplot(data=clave_df, aes(x=Sample, y=Abundance, fill=ta7_ab))+ 
    geom_bar(aes(), stat="identity", position="stack")+
    scale_fill_manual(values = phylum_colors_rel)
#absolute_plot | 
relative_plot
```

```{r}
key_not_key <- c()
for (i in 1:dim(phy_maiz@otu_table@.Data)[1]){
  
  if (is.element( row.names(phy_maiz@otu_table@.Data)[i]   , row.names(phy_maiz_key@otu_table@.Data)  )   ){
    key_not_key <- c(key_not_key , "key")
  } else {
    key_not_key <- c(key_not_key , "not_key")
  
    }
}
```

```{r}

keys_vs_median <- phy_maiz@otu_table@.Data[which(key_not_key == "key") , ]
medians <- c()
means <- c()
for (i in 1:dim(phy_maiz@otu_table@.Data)[2]){
  
  medians <- c(medians , median(phy_maiz@otu_table@.Data[, i]))
  
  }

keys_vs_median <- rbind(keys_vs_median , medians)
#keys_vs_median <- rbind(keys_vs_median , means)
keys_vs_median <-  psmelt(otu_table(keys_vs_median  , taxa_are_rows = TRUE))
```









```{r}
ggplot(keys_vs_median , aes(x = Sample , y = Abundance , col = OTU , group = OTU))  + geom_line( ) + facet_wrap(~OTU)
#ggplot(, aes(x = Sample , y = Abundance , color = OTU)) 
```

```{r}






key_vs_no_key <-  psmelt(phy_maiz@otu_table)

key_not_key <- c()
for (i in 1:dim(key_vs_no_key)[1]){
  
  if (is.element( key_vs_no_key[i, "OTU"]   , row.names(phy_maiz_key@otu_table@.Data)  )   ){
    key_not_key <- c(key_not_key , "key")
  } else {
    key_not_key <- c(key_not_key , "not_key")
  
    }
}
key_vs_no_key <- cbind(key_vs_no_key , key_not_key)
#medians <- psmelt(otu_table(as.data.frame(medians) , taxa_are_rows = FALSE))
```



```{r}
ggplot(key_vs_no_key , aes(x = Sample , y= Abundance , color = key_not_key , group = key_not_key))+ #geom_dotplot(binaxis='y', stackdir='center') +
  stat_summary(fun.data="mean_se", fun.args = list(mult=1), 
                 geom="crossbar", width=0.5) +
 stat_summary(fun="median", fun.args = list(mult=1), 
                 geom="line", width=0.5)
#  geom_errorbar(aes(ymin=Abundance-sd, ymax=Abundance+sd), width=.2,
 #                position=position_dodge(0.05))
```























```{r}
data_maiz <- read.table("../../data/tables/table.from_maiz.txt" , row.names = 1, header = TRUE , sep = "" )

medias <- c()
medianas <- c()
for (i in 1:dim(data_maiz)[2]){
  medias <- c(medias , mean(data_maiz[ , i]))
  medianas <- c(medianas , median(data_maiz[ , i]))
  }

data_maiz <- data_maiz[ row.names(otus_centrales[[3]]),]
data_maiz <- rbind(medias , data_maiz)
data_maiz <- rbind(medianas , data_maiz)

row.names(data_maiz) <- c("Medianas" , "Medias" , row.names(otus_centrales[[3]]))
head(data_maiz)
```

```{r}
V(redes[[3]])$label <- NA
V(redes[[3]])$size <- 3
plot(redes[[3]])
```
```{r}
nombres_otus_centrales_maiz <- as.character(otus_centrales[[3]][,"nodos"])
red_otus_centrales_maiz <- induced_subgraph(redes[[3]], nombres_otus_centrales_maiz  ,"auto")
print( otus_centrales[[3]][,"nodos"])

print(row.names(otus_centrales[[3]]))
print(degree( red_otus_centrales_maiz))
#V(net_work_centrales)$label <- tax_centrales[,"Rank7"]
```



```{r}
plot(red_otus_centrales_maiz)

```


```{r}
otus_no_centrales_maiz <- names(V(redes[[3]]))[ is.element(names(V(redes[[3]])), nombres_otus_centrales_maiz) == FALSE]
red_otus_no_centrales_maiz <- induced_subgraph(redes[[3]] , otus_no_centrales_maiz , "auto" )
plot(red_otus_no_centrales_maiz)
```






























```{r}
library(phyloseq)
familias <- union(unique(rizosferas[[1]][,"V7"]), unique(rizosferas[[2]][,"V7"]))
familias <- union(familias , unique(rizosferas[[3]][,"V7"]))

muestras_totales <- union(colnames(data_chile), colnames(data_maiz))
muestras_totales <- union(muestras_totales , colnames(data_maiz))

familia_x_muestras <- matrix(nrow = length(familias) , ncol = length(muestras_totales))

x <- 0
for (i in 1:3) {
  for (j in 1:dim(data[[i]])[2]){
    x <- x + 1
    for (k in 1:length(familias)){
      if (is.element(familias[k], rizosferas[[i]][, "V7"])){
        familia_x_muestras[k, x] <- sum(data[[i]][, j][which( rizosferas[[i]][, "V7"] == familias[k]  ) ])
      } else { familia_x_muestras[k, x] <- 0 }
    }
  }
}

row.names(familia_x_muestras) <- familias
colnames(familia_x_muestras) <- muestras_totales

familia_x_muestras <- data.frame(familia_x_muestras)
familia_x_muestras <- otu_table(familia_x_muestras , taxa_are_rows = TRUE)
```

```{r}
planta <- c()
for (i in 1:length(muestras_totales)){
  if (is.element(muestras_totales[i], colnames(data[[1]]) )){ planta <- c(planta , "chile") }
  else {if (is.element(muestras_totales[i], colnames(data[[2]]) )){ planta <- c(planta , "maiz") } 
  else {planta <- c(planta , "maiz") }}
}
plant_sample <- data.frame( ID = muestras_totales , Planta = planta , row.names = muestras_totales )
plant_sample <- sample_data(plant_sample)
```

```{r}
phy_totales <- phyloseq( otu_table = familia_x_muestras , sample_data = plant_sample )
phy_totales <- transform_sample_counts(phy_totales , function(x) x / sum(x) )
meta_ord_clave <- ordinate( phy_totales , method = "PCoA", distance = "bray")
plot_sample <- plot_ordination(physeq = phy_totales , ordination = meta_ord_clave , color = "Planta")
plot_sample
#ggsave("./plot_pcoa_otus2" , plot_sample , device = "png")
```

```{r}
familias_clave <- union(unique(rizosferas[[1]][row.names(otus_centrales[[1]]) ,"V7"]), unique(rizosferas[[2]][ row.names(otus_centrales[[2]]) ,"V7"]))
familias_clave <- union(familias_clave , unique(rizosferas[[3]][row.names(otus_centrales[[2]])  ,"V7"]))
#familias_clave <- unique(rizosferas[[3]][row.names(otus_centrales[[3]]) ,"V7"])

familia_clave_x_muestras <- matrix(nrow = length(familias_clave) , ncol = length(muestras_totales))

x <- 0
for (i in 1:3) {
  for (j in 1:dim(data[[i]])[2]){
    x <- x + 1
    for (k in 1:length(familias_clave)){
      if (is.element(familias_clave[k], rizosferas[[i]][, "V7"])){
        familia_clave_x_muestras[k, x] <- sum(data[[i]][, j][which( rizosferas[[i]][, "V7"] == familias[k]  ) ])
      } else { familia_clave_x_muestras[k, x] <- 0 }
    }
  }
}

row.names(familia_clave_x_muestras) <- familias_clave
colnames(familia_clave_x_muestras) <- muestras_totales

familia_clave_x_muestras <- data.frame(familia_clave_x_muestras)
familia_clave_x_muestras <- otu_table(familia_clave_x_muestras , taxa_are_rows = TRUE)
```




```{r}
phy_totales <- phyloseq( otu_table = familia_clave_x_muestras , sample_data = plant_sample )
phy_totales <- transform_sample_counts(phy_totales , function(x) x / sum(x) )
meta_ord_clave <- ordinate( phy_totales , method = "PCoA", distance = "bray")
plot_sample <- plot_ordination(physeq = phy_totales , ordination = meta_ord_clave , color = "Planta")
plot_sample
#ggsave("./plot_pcoa_otus2" , plot_sample , device = "png")
```









```{r}
library(phyloseq)
familias <- union(unique(rizosferas[[1]][,"V8"]), unique(rizosferas[[2]][,"V8"]))
familias <- union(familias , unique(rizosferas[[3]][,"V8"]))

muestras_totales <- union(colnames(data_chile), colnames(data_maiz))
muestras_totales <- union(muestras_totales , colnames(data_maiz))

familia_x_muestras <- matrix(nrow = length(familias) , ncol = length(muestras_totales))

x <- 0
for (i in 1:3) {
  for (j in 1:dim(data[[i]])[2]){
    x <- x + 1
    for (k in 1:length(familias)){
      if (is.element(familias[k], rizosferas[[i]][, "V8"])){
        familia_x_muestras[k, x] <- sum(data[[i]][, j][which( rizosferas[[i]][, "V8"] == familias[k]  ) ])
      } else { familia_x_muestras[k, x] <- 0 }
    }
  }
}

row.names(familia_x_muestras) <- familias
colnames(familia_x_muestras) <- muestras_totales

familia_x_muestras <- data.frame(familia_x_muestras)
familia_x_muestras <- otu_table(familia_x_muestras , taxa_are_rows = TRUE)
```

```{r}
planta <- c()
for (i in 1:length(muestras_totales)){
  if (is.element(muestras_totales[i], colnames(data[[1]]) )){ planta <- c(planta , "chile") }
  else {if (is.element(muestras_totales[i], colnames(data[[2]]) )){ planta <- c(planta , "maiz") } 
  else {planta <- c(planta , "maiz") }}
}
plant_sample <- data.frame( ID = muestras_totales , Planta = planta , row.names = muestras_totales )
plant_sample <- sample_data(plant_sample)
```

```{r}
phy_totales <- phyloseq( otu_table = familia_x_muestras , sample_data = plant_sample )
phy_totales <- transform_sample_counts(phy_totales , function(x) x / sum(x) )
meta_ord_clave <- ordinate( phy_totales , method = "PCoA", distance = "bray")
plot_sample <- plot_ordination(physeq = phy_totales , ordination = meta_ord_clave , color = "Planta")
plot_sample
#ggsave("./plot_pcoa_otus2" , plot_sample , device = "png")
```

```{r}
familias_clave <- union(unique(rizosferas[[1]][row.names(otus_centrales[[1]]) ,"V8"]), unique(rizosferas[[2]][ row.names(otus_centrales[[2]]) ,"V8"]))
familias_clave <- union(familias_clave , unique(rizosferas[[3]][row.names(otus_centrales[[2]])  ,"V8"]))
#familias_clave <- unique(rizosferas[[3]][row.names(otus_centrales[[3]]) ,"V7"])

familia_clave_x_muestras <- matrix(nrow = length(familias_clave) , ncol = length(muestras_totales))

x <- 0
for (i in 1:3) {
  for (j in 1:dim(data[[i]])[2]){
    x <- x + 1
    for (k in 1:length(familias_clave)){
      if (is.element(familias_clave[k], rizosferas[[i]][, "V8"])){
        familia_clave_x_muestras[k, x] <- sum(data[[i]][, j][which( rizosferas[[i]][, "V8"] == familias[k]  ) ])
      } else { familia_clave_x_muestras[k, x] <- 0 }
    }
  }
}

row.names(familia_clave_x_muestras) <- familias_clave
colnames(familia_clave_x_muestras) <- muestras_totales

familia_clave_x_muestras <- data.frame(familia_clave_x_muestras)
familia_clave_x_muestras <- otu_table(familia_clave_x_muestras , taxa_are_rows = TRUE)
```




```{r}
phy_totales <- phyloseq( otu_table = familia_clave_x_muestras , sample_data = plant_sample )
phy_totales <- transform_sample_counts(phy_totales , function(x) x / sum(x) )
meta_ord_clave <- ordinate( phy_totales , method = "PCoA", distance = "bray")
plot_sample <- plot_ordination(physeq = phy_totales , ordination = meta_ord_clave , color = "Planta")
plot_sample
#ggsave("./plot_pcoa_otus2" , plot_sample , device = "png")
```
```{r}
library(phyloseq)
familias <- union(unique(rizosferas[[1]][,"V7"]), unique(rizosferas[[2]][,"V9"]))
familias <- union(familias , unique(rizosferas[[3]][,"V9"]))

muestras_totales <- union(colnames(data_chile), colnames(data_tomate))
muestras_totales <- union(muestras_totales , colnames(data_maiz))

familia_x_muestras <- matrix(nrow = length(familias) , ncol = length(muestras_totales))

x <- 0
for (i in 1:3) {
  for (j in 1:dim(data[[i]])[2]){
    x <- x + 1
    for (k in 1:length(familias)){
      if (is.element(familias[k], rizosferas[[i]][, "V9"])){
        familia_x_muestras[k, x] <- sum(data[[i]][, j][which( rizosferas[[i]][, "V9"] == familias[k]  ) ])
      } else { familia_x_muestras[k, x] <- 0 }
    }
  }
}

row.names(familia_x_muestras) <- familias
colnames(familia_x_muestras) <- muestras_totales

familia_x_muestras <- data.frame(familia_x_muestras)
familia_x_muestras <- otu_table(familia_x_muestras , taxa_are_rows = TRUE)
```

```{r}
planta <- c()
for (i in 1:length(muestras_totales)){
  if (is.element(muestras_totales[i], colnames(data[[1]]) )){ planta <- c(planta , "chile") }
  else {if (is.element(muestras_totales[i], colnames(data[[2]]) )){ planta <- c(planta , "tomate") } 
  else {planta <- c(planta , "maiz") }}
}
plant_sample <- data.frame( ID = muestras_totales , Planta = planta , row.names = muestras_totales )
plant_sample <- sample_data(plant_sample)
```

```{r}
phy_totales <- phyloseq( otu_table = familia_x_muestras , sample_data = plant_sample )
phy_totales <- transform_sample_counts(phy_totales , function(x) x / sum(x) )
meta_ord_clave <- ordinate( phy_totales , method = "PCoA", distance = "bray")
plot_sample <- plot_ordination(physeq = phy_totales , ordination = meta_ord_clave , color = "Planta")
plot_sample
#ggsave("./plot_pcoa_otus2" , plot_sample , device = "png")
```

```{r}
familias_clave <- union(unique(rizosferas[[1]][row.names(otus_centrales[[1]]) ,"V9"]), unique(rizosferas[[2]][ row.names(otus_centrales[[2]]) ,"V9"]))
familias_clave <- union(familias_clave , unique(rizosferas[[3]][row.names(otus_centrales[[2]])  ,"V9"]))
#familias_clave <- unique(rizosferas[[3]][row.names(otus_centrales[[3]]) ,"V7"])

familia_clave_x_muestras <- matrix(nrow = length(familias_clave) , ncol = length(muestras_totales))

x <- 0
for (i in 1:3) {
  for (j in 1:dim(data[[i]])[2]){
    x <- x + 1
    for (k in 1:length(familias_clave)){
      if (is.element(familias_clave[k], rizosferas[[i]][, "V9"])){
        familia_clave_x_muestras[k, x] <- sum(data[[i]][, j][which( rizosferas[[i]][, "V9"] == familias[k]  ) ])
      } else { familia_clave_x_muestras[k, x] <- 0 }
    }
  }
}

row.names(familia_clave_x_muestras) <- familias_clave
colnames(familia_clave_x_muestras) <- muestras_totales

familia_clave_x_muestras <- data.frame(familia_clave_x_muestras)
familia_clave_x_muestras <- otu_table(familia_clave_x_muestras , taxa_are_rows = TRUE)
```




```{r}
phy_totales <- phyloseq( otu_table = familia_clave_x_muestras , sample_data = plant_sample )
phy_totales <- transform_sample_counts(phy_totales , function(x) x / sum(x) )
meta_ord_clave <- ordinate( phy_totales , method = "PCoA", distance = "bray")
plot_sample <- plot_ordination(physeq = phy_totales , ordination = meta_ord_clave , color = "Planta")
plot_sample
#ggsave("./plot_pcoa_otus2" , plot_sample , device = "png")
```


```{r}
otus_totales <- union(row.names(data_chile), row.names(data_tomate) )
otus_totales <- union(otus_totales , row.names(data_maiz))
muestras_totales <- union(colnames(data_chile), colnames(data_tomate))
muestras_totales <- union(muestras_totales , colnames(data_maiz))
data_totales <- matrix( nrow = length(otus_totales) , ncol = length(muestras_totales) )

x <- 0
for (i in 1:3) {
  for (j in 1:dim(data[[i]])[2]){
    x <- x + 1
    for (k in 1:length(otus_totales)){
      if (is.element(otus_totales[k], row.names(data[[i]]) )){
        data_totales[k, x] <- data[[i]][otus_totales[k], j]
      } else { data_totales[k, x] <- 0 }
    }
  }
}
```


```{r}
row.names(data_totales) <- otus_totales
colnames(data_totales) <- muestras_totales
data_totales <- data.frame(data_totales)
data_totales <- otu_table(data_totales , taxa_are_rows = TRUE)
```

```{r}
planta <- c()
for (i in 1:length(muestras_totales)){
  if (is.element(muestras_totales[i], colnames(data[[1]]) )){ planta <- c(planta , "chile") }
  else {if (is.element(muestras_totales[i], colnames(data[[2]]) )){ planta <- c(planta , "tomate") } 
  else {planta <- c(planta , "maiz") }}
}
plant_sample <- data.frame( ID = muestras_totales , Planta = planta , row.names = muestras_totales )
plant_sample <- sample_data(plant_sample)
```


```{r}
phy_totales <- phyloseq( otu_table = data_totales , sample_data = plant_sample )
meta_ord_clave <- ordinate( phy_totales , method = "NMDS", distance = "bray")
plot_sample <- plot_ordination(physeq = phy_totales , ordination = meta_ord_clave , color = "Planta")
plot_sample
#ggsave("./plot_pcoa_otus2" , plot_sample , device = "png")
```
```{r}
phy_totales <- phyloseq( otu_table = data_totales , sample_data = plant_sample )
meta_ord_clave <- ordinate( phy_totales , method = "PCoA", distance = "bray")
plot_sample <- plot_ordination(physeq = phy_totales , ordination = meta_ord_clave , color = "Planta")
plot_sample
#ggsave("./plot_pcoa_otus2" , plot_sample , device = "png")
```



```{r}
otus_centrales_chile <- read.csv("../central_otus/reporte_chile1.csv" , row.names = 1)
otus_centrales_tomate <- read.csv("../central_otus/reporte_tomate1.csv" , row.names = 1)
otus_centrales_maiz <- read.csv("../central_otus/reporte_maiz1.csv" , row.names = 1)

otus_centrales <- list(otus_centrales_chile , otus_centrales_tomate , otus_centrales_maiz)
```

```{r}

for (i in 1:3){
  otus_centrales[[i]] <- data[[i]][ intersect( row.names(otus_centrales[[i]]), row.names(data[[i]]) ) ,]
  }
```

```{r}
clave_totales <- union(row.names(otus_centrales[[1]]), row.names(otus_centrales[[2]]) )
clave_totales <- union(clave_totales , row.names(otus_centrales[[3]]))
#muestras_totales <- union(colnames(data_chile), colnames(data_tomate))
#muestras_totales <- union(muestras_totales , colnames(data_maiz))
data_clave_totales <- matrix( nrow = length(clave_totales) , ncol = length(muestras_totales) )

x <- 0
for (i in 1:3) {
  for (j in 1:dim(otus_centrales[[i]])[2]){
    x <- x + 1
    for (k in 1:length(clave_totales)){
      if (is.element(clave_totales[k], row.names(otus_centrales[[i]]) )){
        data_clave_totales[k, x] <- otus_centrales[[i]][clave_totales[k], j]
      } else { data_clave_totales[k, x] <- 0 }
    }
  }
}
```

```{r}
row.names(data_clave_totales) <- clave_totales
colnames(data_clave_totales) <- muestras_totales
data_clave_totales <- data.frame(data_clave_totales)
data_clave_totales <- otu_table(data_clave_totales , taxa_are_rows = TRUE)
```

```{r}
phy_claves <- phyloseq( otu_table = data_clave_totales , sample_data = plant_sample )
pht_claves <- transform_sample_counts(phy_claves , function(x) x / sum(x) )
meta_ord_clave <- ordinate( phy_claves , method = "NMDS", distance = "bray")
plot_sample <- plot_ordination(physeq = phy_claves , ordination = meta_ord_clave , color = "Planta")
plot_sample
#ggsave("./plot_pcoa_otus2" , plot_sample , device = "png")
```
```{r}
#phy_claves <- phyloseq( otu_table = data_clave_totales , sample_data = plant_sample )
meta_ord_clave <- ordinate( phy_claves , method = "PCoA", distance = "bray")
plot_sample <- plot_ordination(physeq = phy_claves , ordination = meta_ord_clave , color = "Planta")
plot_sample
#ggsave("./plot_pcoa_otus2" , plot_sample , device = "png")
```


